titanic.py

```diff
@@ -93,12 +93,46 @@ def maximum_energy_delta(bqm):
                      for j in bqm.iter_neighbors(i))
                for i in bqm.iter_variables())
 
-def run_demo():
+def mutual_information_bqm(dataset, features):
+    """Build a QUBO that maximizes MI between survival and a subset of features"""
+    variables = ((feature, -mutual_information(prob(dataset[['survived', feature]].values), 1))
+                 for feature in features)
+    interactions = ((f0, f1, -conditional_mutual_information(prob(dataset[['survived', f0, f1]].values), 1, 2))
+                    for f0, f1 in itertools.permutations(features, 2))
+    return dimod.BinaryQuadraticModel(variables, interactions, 0, dimod.BINARY)
+
+def add_combination_penalty(bqm, k, penalty):
+    """Create a new BQM with an addition k-combination penalty"""
+    kbqm = dimod.generators.combinations(bqm.variables, k, strength=penalty)
+    kbqm.update(bqm)
+    return kbqm
+
+def mutual_information_feature_selection(dataset, features):
+    """Run the MIFS algoeith on D-Wave"""
+    
+    # Set up a QPU sampler with a fully-connected graph of all the variables
+    sampler = DWaveCliqueSampler()
+
+    # For each number of features, k, penalize selection of fewer or more features
+    selected_features = np.zeros((len(features), len(features)))
+
+    bqm = mutual_information_bqm(dataset, features)
+
+    # This ensures that the ground state will satisfy the constraints.
+    penalty = maximum_energy_delta(bqm)
+
+    for k in range(1, len(features) + 1):
+        kbqm = add_combination_penalty(bqm, k, penalty)
+        sample = sampler.sample(kbqm,
+                                label='Example - MI Feature Selection',
+                                num_reads=10000).first.sample
```

Comment:
Just noticed that large number of samples: we rarely use the maximum of the range, is it really needed here?

Reply:
That was there before I just moved the code into a function

Reply:
Yes, that is what I meant by "Just noticed", that I hadn't seen it previously although it was there. It's a general comment--we should look at the returned solutions and see if it makes sense to reduce the number of requested samples to something more typical (1000, 5000) or if it's needed in this case

Reply:
Should I change that in this PR? What to?

Reply:
I am not sure we have an independent basis to judge what makes sense for this example

Reply:
For k=4 about 2% of the samples are the lowest-energy state found:
![image](https://user-images.githubusercontent.com/34041130/109827571-b42abf80-7bf0-11eb-86e8-3c3c7c2b1e65.png)

Reply:
For the few runs I just did, that seems about the worst case, and I'm seeing the same lowest energy in HSS, so I suspect that 5000 reads is likely fine too.

Reply:
done